### PR TITLE
[apex] ApexCSRF changes

### PR DESCRIPTION
**Pull request by adangel**
**PR Description:**

This is a follow-up on #2064 

* Moves the rule from security to errorprone (and leaves a deprecated rule-ref behind in security for compatibility). Note, this also required to move the Helper class.
* Extends the description of the rule
* Fixes false-negatives with (static) initializers

About the "init" method: It might be, that it is the action attribute, e.g.:

`<apex:page controller="CSRF_Demo" sidebar="false" action="{!init}" tabStyle="CSRF_Demo__tab">`

On https://trailhead.salesforce.com/en/content/learn/modules/secdev_application_logic_vulnerabilities/secdev_app_logic_csrf there is also one important sentence:

> To prevent this issue [CSRF], modify your application logic to remove any state-changing 
> operations from the apex:page action handlers. This ensures that all requests are protected by 
> the default Salesforce CSRF protections.

So, the default Salesforce CSRF protections only work, if there are no DML statements in any initializing code of the page controller.

@rsoesemann  Wdyt?

Before submitting a PR, please check that:
 - [x] The PR is submitted against `master`. The PMD team will merge back to support branches as needed.
 - [x] `./mvnw clean verify` passes. This will [build](https://github.com/pmd/pmd/blob/master/BUILDING.md) and test PMD, execute PMD and checkstyle rules. [Check this for more info](https://github.com/pmd/pmd/blob/master/CONTRIBUTING.md#code-style)

### CODE REVIEW

**Comment by pmd-test** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="true">This changeset introduces 0 new violations, 0 new errors and 0 new configuration errors,<br />
removes 0 violations, 0 errors and 0 configuration errors.<br />
<a href="https://chunk.io/pmd/3a827fb64ac64b37a0d87cd410642df3/diff/index.html">Full report</a></td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: pmd-apex/src/main/java/net/sourceforge/pmd/lang/apex/rule/errorprone/ApexCSRFRule.java [hunk 1]
```diff
@@ -17,6 +19,7 @@
  */
 public class ApexCSRFRule extends AbstractApexRule {
     public static final String INIT = "init";
+    private static final String STATIC_INITIALIZER = "<clinit>";
```
Comment (rsoesemann) [inline pmd-apex/src/main/java/net/sourceforge/pmd/lang/apex/rule/errorprone/ApexCSRFRule.java]: Sorry if I continue to ask this question but we must get rid of this CodeClimate stuff. I, we, the Salesforce world and you the main maintainer dont want provider specific stuff in PMD. When will it finally be removed?

File: pmd-apex/src/main/java/net/sourceforge/pmd/lang/apex/rule/errorprone/ApexCSRFRule.java [hunk 2]
```diff
@@ -41,6 +44,17 @@ public Object visit(ASTMethod node, Object data) {
         return data;
     }
 
+    @Override
+    public Object visit(ASTBlockStatement node, Object data) {
+        if (node.jjtGetParent() instanceof ASTUserClass) {
+            // initializer block
```
Comment (rsoesemann) [inline pmd-apex/src/main/java/net/sourceforge/pmd/lang/apex/rule/errorprone/ApexCSRFRule.java]: Unneeded comment

File: pmd-apex/src/main/java/net/sourceforge/pmd/lang/apex/rule/errorprone/ApexCSRFRule.java [hunk 3]
```diff
@@ -54,7 +68,7 @@ private void checkForCSRF(ASTMethod node, Object data) {
         }
 
         String name = node.getImage();
-        if (name.equalsIgnoreCase(INIT)) {
+        if (INIT.equalsIgnoreCase(name) || STATIC_INITIALIZER.equals(name)) {
```
Comment (rsoesemann) [inline pmd-apex/src/main/java/net/sourceforge/pmd/lang/apex/rule/errorprone/ApexCSRFRule.java]: I'm not a java expert but I would combine those two ifs into one. If it gets unreadable then I would extract the outer two tests into a method

File: pmd-apex/src/main/java/net/sourceforge/pmd/lang/apex/rule/internal/Helper.java [hunk 4]
```diff
@@ -0,0 +1,230 @@
...
+            }
+        }
+
+        return false;
+    }
+
+    public static String getFQVariableName(final ASTVariableExpression variable) {
+        final ASTReferenceExpression ref = variable.getFirstChildOfType(ASTReferenceExpression.class);
+        String objectName = "";
+        if (ref != null) {
```
Comment (rsoesemann) [inline pmd-apex/src/main/java/net/sourceforge/pmd/lang/apex/rule/internal/Helper.java]: Combine ifs here?

File: pmd-apex/src/main/java/net/sourceforge/pmd/lang/apex/rule/internal/Helper.java [hunk 5]
```diff
@@ -0,0 +1,230 @@
...
+                .append(variable.getNode().getDefiningType().getApexName()).append(":")
+                .append(variable.getType());
+        return sb.toString();
+    }
+
+    public static boolean isSystemLevelClass(ASTUserClass node) {
+        List<TypeRef> interfaces = node.getNode().getDefiningType().getCodeUnitDetails().getInterfaceTypeRefs();
+
+        for (TypeRef intObject : interfaces) {
+            if (isWhitelisted(intObject.getNames())) {
```
Comment (rsoesemann) [inline pmd-apex/src/main/java/net/sourceforge/pmd/lang/apex/rule/internal/Helper.java]: Why an if and then an extra return true? I'd just:

`return isWhitelisted(...);`

File: pmd-apex/src/main/java/net/sourceforge/pmd/lang/apex/rule/internal/Helper.java [hunk 6]
```diff
@@ -0,0 +1,230 @@
...
+            }
+        }
+
+        return false;
+    }
+
+    private static boolean isWhitelisted(List<Identifier> ids) {
+        StringBuffer sb = new StringBuffer();
+
+        for (int i = 0; i < ids.size(); i++) {
```
Comment (rsoesemann) [inline pmd-apex/src/main/java/net/sourceforge/pmd/lang/apex/rule/internal/Helper.java]: Isn't the a more redable for each in Java?

File: pmd-apex/src/main/java/net/sourceforge/pmd/lang/apex/rule/internal/Helper.java [hunk 7]
```diff
@@ -0,0 +1,230 @@
...
+            sb.append(ids.get(i).getValue());
+
+            if (i != ids.size() - 1) {
+                sb.append(".");
+            }
+        }
+
+        switch (sb.toString().toLowerCase(Locale.ROOT)) {
+        case "queueable":
```
Comment (rsoesemann) [inline pmd-apex/src/main/java/net/sourceforge/pmd/lang/apex/rule/internal/Helper.java]: Indentation?

File: pmd-core/src/test/java/net/sourceforge/pmd/RuleSetFactoryTest.java [hunk 8]
```diff
@@ -349,6 +349,42 @@ public void testRuleSetReferencesRulesetWithAExcludedNonExistingRule() throws Ex
                     "WARNING: Unable to exclude rules [NonExistingRule] from ruleset reference rulesets/dummy/basic.xml; perhaps the rule name is mispelled or the rule doesn't exist anymore?"));
     }
 
+    /**
+     * When a custom ruleset references a ruleset that only contains deprecated rules, then this ruleset itself is
+     * considered deprecated and the user should get a deprecation warning for the ruleset.
+     */
+    @Test
+    public void testRuleSetReferencesDeprecatedRuleset() throws Exception {
```
Comment (rsoesemann) [inline pmd-core/src/test/java/net/sourceforge/pmd/RuleSetFactoryTest.java]: I think test methods get more reddable (scannable) when not every method is prefixed with test.

File: pmd-apex/src/main/java/net/sourceforge/pmd/lang/apex/rule/errorprone/ApexCSRFRule.java (hunk 1 above)
Comment (adangel) [reply pmd-apex/src/main/java/net/sourceforge/pmd/lang/apex/rule/errorprone/ApexCSRFRule.java]: It's still the same plan: We've already deprecated the properties in PMD 6 and will remove them (probably have them already in the branch) in PMD 7.

File: pmd-apex/src/main/java/net/sourceforge/pmd/lang/apex/rule/internal/Helper.java (hunk 5 above)
Comment (adangel) [reply pmd-apex/src/main/java/net/sourceforge/pmd/lang/apex/rule/internal/Helper.java]: That would change the logic: We would then only consider the first interface... (we are in a loop here)

File: pmd-apex/src/main/java/net/sourceforge/pmd/lang/apex/rule/internal/Helper.java (hunk 6 above)
Comment (adangel) [reply pmd-apex/src/main/java/net/sourceforge/pmd/lang/apex/rule/internal/Helper.java]: Not if you need access to the index. What's done here is basically joining the identifiers with ".". With java8 (and I think, we are here in Apex already with java8) we can use Streams to do that.

File: pmd-apex/src/main/java/net/sourceforge/pmd/lang/apex/rule/internal/Helper.java (hunk 7 above)
Comment (adangel) [reply pmd-apex/src/main/java/net/sourceforge/pmd/lang/apex/rule/internal/Helper.java]: Hm... we seem to have a different (checkstyle) rule active for indenting switches... the way it is currently is correct (at least we enforce it so through checkstyle)

**Comment by adangel** [comment]
Thanks for having a look at it. I think, that's now ready.

**Pull request merged**